OpenRA.Mods.Common/Traits/Player/TimeLimitManager.cs

```diff
@@ -0,0 +1,133 @@
+		[Desc("Display order for the time limit dropdown in the lobby.")]
+		public readonly int TimeLimitDisplayOrder = 0;
+
+		IEnumerable<LobbyOption> ILobbyOptions.LobbyOptions(Ruleset rules)
+		{
+			var timelimits = TimeLimitOptions.ToDictionary(c => c.ToString(), c =>
+			{
+				if (c == 0)
+					return "No limit";
+				else
+					return c.ToString() + " minute{0}".F(c > 1 ? "s" : null);
+			});
+
+			yield return new LobbyOption("timelimit", TimeLimitLabel, TimeLimitDescription, true, TimeLimitDisplayOrder,
+				new ReadOnlyDictionary<string, string>(timelimits), TimeLimitDefault.ToString(), TimeLimitLocked);
+		}
+
+		public object Create(ActorInitializer init) { return new TimeLimitManager(init.Self, this); }
+	}
+
+	public class TimeLimitManager : INotifyTimeLimit, ITick
+	{
+		TimeLimitManagerInfo info;
+		public readonly int TimeLimit;
+
+		public TimeLimitManager(Actor self, TimeLimitManagerInfo info)
+		{
+			this.info = info;
+
+			var tl = self.World.LobbyInfo.GlobalSettings.OptionOrDefault("timelimit", info.TimeLimitDefault.ToString());
+			if (!int.TryParse(tl, out TimeLimit))
+				TimeLimit = info.TimeLimitDefault;
+
+			// Convert from minutes to ticks
+			TimeLimit *= 60 * (1000 / self.World.Timestep);
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (TimeLimit <= 0)
+				return;
+
+			var ticksPerSecond = 1000 / self.World.Timestep;
+			var ticksRemaining = TimeLimit - self.World.WorldTick;
+
+			if (ticksRemaining > 0 && info.TimeLimitWarnings.Keys.All(m => ticksRemaining != m * 60 * ticksPerSecond))
+				return;
+
+			var ntls = self.TraitsImplementing<INotifyTimeLimit>();
+
+			if (ticksRemaining == 0)
+			{
+				foreach (var ntl in ntls)
+					ntl.NotifyTimerExpired(self);
+
+				return;
+			}
+
+			foreach (var ntl in ntls)
+				foreach (var t in info.TimeLimitWarnings.Keys)
+					if (ticksRemaining == t * 60 * ticksPerSecond)
+						ntl.NotifyTimeRemaining(self, t, info.TimeLimitWarnings[t]);
+		}
+
+		void INotifyTimeLimit.NotifyTimeRemaining(Actor self, int minutesRemaining, string audioNotification)
+		{
+			if (self.Owner == self.World.RenderPlayer || self.World.RenderPlayer == null)
+			{
+				Game.AddChatLine(Color.White, "Battlefield Control", "Match timer: {0} minute{1} remaining.".F(minutesRemaining, minutesRemaining > 1 ? "s" : null));
+				Game.Sound.PlayNotification(self.World.Map.Rules, self.Owner, "Speech", audioNotification, self.Owner.Faction.InternalName);
+			}
+		}
+
+		void INotifyTimeLimit.NotifyTimerExpired(Actor self)
+		{
+			if (self.Owner == self.World.RenderPlayer || self.World.RenderPlayer == null)
+				Game.AddChatLine(Color.White, "Battlefield Control", "Match timer has expired.");
+
+			// The 'updateWinners' flag is set to false here to ensure that all players lose when the time is up.
+			// Without it, the last player in the loop would win by default.
+			foreach (var p in self.World.Players.Where(p => p.HasObjectives))
+				p.PlayerActor.Trait<MissionObjectives>().ForceDefeat(p, false);
```

Comment:
IMO this feature would be much better if it declared the player with the highest score the winner.

Reply:
Thanks for mentioning that. I wasn't sure which way to go about this. From my POV there were three options: 1) everyone loses, 2) add a `WinState.Draw`, 3) let the victory conditions decide. I only didn't go with option 3) because I feel that our game score is still somewhat half-assed (although there's usually a strong correlation between scores and winner) and I'm not sure how players would feel about it.

Reply:
If this prompted discussion and improvements to the score system, then all the better IMO!